Task concurrency management in K8s Task Scheduling extension

YongGang opened the pull request:
### Description

This PR enhances task management capabilities by introducing two features: Task Labeling and Task Scheduling, utilizing the available task slots in the Kubernetes Task Runner. These features are designed to manage a diverse array of task types more efficiently, particularly addressing the complexities of Multi-Stage Query (MSQ) tasks that perform varied functions. These enhancements are designed to provide granular control over task execution and extend the capabilities for monitoring and analysis of task performance.

**Task Labeling**
We've integrated a Task Label Concept to explicitly clarify each task's purpose. This concept allows for distinguishing between tasks that, although of the same type, serve different operational roles within Druid's ecosystem. Task labeling enables more precise monitoring, reporting, and management of tasks based on their designated labels.
(a separate PR created for this change for easier review https://github.com/apache/druid/pull/16041)

**Task Scheduling**
Enhancements to Task Scheduling leverage task labels to optimize the allocation of Kubernetes task slots. This ensures that tasks are scheduled according to their operational priority and resource requirements, leading to improved cluster capacity management. The example configuration demonstrates the application of this feature:
1. limit the admin tasks (`kill,compact`) to use max 10% of the cluster capacity.
2. limit querying from deep storage task (`msq_export_durable_storage`) to use max 20% of the cluster capacity.
3. Reserve 60% of cluster capacity for (critical) ingestion tasks (`index_parallel,msq_index_batch`)
```
druid.indexer.runner.taskLanes=[
  {"labels":"kill,compact","capacityRatio":"0.1","policy":"MAX"},
  {"labels":"msq_export_durable_storage","capacityRatio":"0.2","policy":"MAX"},
  {"labels":"index_parallel,msq_index_batch","capacityRatio":"0.6","policy":"RESERVE"}
]
```
<!--
In each section, please describe design decisions made, including:
 - Choice of algorithms
 - Behavioral aspects. What configuration values are acceptable? How are corner cases and error conditions handled, such as when there are insufficient resources?
 - Class organization and design (how the logic is split between classes, inheritance, composition, design patterns)
 - Method organization and design (how the logic is split between methods, parameters and return types)
 - Naming (class, method, API, configuration, HTTP endpoint, names of emitted metrics)
-->


<!-- It's good to describe an alternative design (or mention an alternative name) for every design (or naming) decision point and compare the alternatives with the designs that you've implemented (or the names you've chosen) to highlight the advantages of the chosen designs and names. -->

<!-- If there was a discussion of the design of the feature implemented in this PR elsewhere (e. g. a "Proposal" issue, any other issue, or a thread in the development mailing list), link to that discussion from this PR description and explain what have changed in your final design compared to your original proposal or the consensus version in the end of the discussion. If something hasn't changed since the original discussion, you can omit a detailed discussion of those aspects of the design here, perhaps apart from brief mentioning for the sake of readability of this PR description. -->

<!-- Some of the aspects mentioned above may be omitted for simple and small changes. -->

#### Release note
<!-- Give your best effort to summarize your changes in a couple of sentences aimed toward Druid users. 

If your change doesn't have end user impact, you can skip this section.

For tips about how to write a good release note, see [Release notes](https://github.com/apache/druid/blob/master/CONTRIBUTING.md#release-notes).

-->


<hr>

##### Key changed/added classes in this PR


<hr>

<!-- Check the items by putting "x" in the brackets for the done things. Not all of these items apply to every PR. Remove the items which are not done or not relevant to the PR. None of the items from the checklist below are strictly necessary, but it would be very helpful if you at least self-review the PR. -->

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

YongGang commented on extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQControllerTask.java:
```diff
@@ -147,6 +149,24 @@ public String getType()
   }
 
+  @Override
+  public String getLabel() {
```
Could @cryptoe @adarshsanjeev please review the proposed labels for MSQ tasks? I want to ensure they accurately reflect the tasks' purposes. Your insights would be greatly appreciated.

adarshsanjeev replied:
The values themselves seem to be correct, and should cover the different use cases for MSQ

arunramani commented on extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/TaskLaneConfig.java:
```diff
@@ -0,0 +1,70 @@
+
+  @JsonCreator
+  public TaskLaneConfig(
+      @JsonProperty("label") String label,
```
The label and config structure feels a little unintuitive. `label` here is actually a comma-separated string of labels. Also the labels themselves are not easy to figure out since they're dynamically generated based on the content of the spec. We should think of a way to make these labels (if we want to keep them at all) be user discoverable.

YongGang replied:
For non-MSQ tasks configuration since label is same as task type I assume user already know that info.
For MSQ tasks configuration, one way is to add more document about how different labels are used in MSQ task so user can find out that but will think about whether there is a better way to do that.

arunramani commented on extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/TaskLaneCapacityPolicy.java:
```diff
@@ -0,0 +1,26 @@
+
+public enum TaskLaneCapacityPolicy
+{
+  MAX,
```
Some description of what these values mean would be great.

YongGang replied:
yup, will do that.

arunramani commented on extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQControllerTask.java:
```diff
@@ -147,6 +149,24 @@ public String getType()
   }
 
+  @Override
+  public String getLabel() {
```
Is laning available only for MSQ tasks for now?

YongGang replied:
It applies to all task types. For non-MSQ task label is same as task type (can refer to the default method added in `Task` interface)

cryptoe commented on indexing-service/src/main/java/org/apache/druid/indexing/common/task/Task.java:
```diff
@@ -134,6 +134,23 @@ default int getPriority()
+   *         Override this method to specify more granular labels tailored to the task's
+   *         specific purpose or operational characteristics.
+   */
+  default String getLabel()
```
It seems weird that we are having an implementation in a contrib extension and not in the core overlord task runner. 
cc @kfaraz 
Wdyt about this approach.

cryptoe commented on extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQWorkerTask.java:
```diff
@@ -116,6 +119,11 @@ public String getType()
 
+  @Override
+  public String getLabel() {
+    return label;
```
You can put the label in context. That way we donot explode the constructor args.

cryptoe commented on extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/TaskLaneCapacityPolicy.java:
```diff
@@ -0,0 +1,42 @@
+ * Task lanes are logical divisions of task execution resources, allowing for more granular control over how tasks are allocated and executed.
+ * This enumeration defines the strategies that can be applied to these lanes in terms of their capacity management.
+ */
+public enum TaskLaneCapacityPolicy
```
I feel these changes should go in core druid. 
Also the AutoCompaction system should use the same mechanisms. No ?

cryptoe commented on extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQControllerTask.java:
```diff
@@ -147,6 +149,24 @@ public String getType()
+  public String getLabel() {
+    if (isIngestion(querySpec)) {
+      if (((DataSourceMSQDestination) querySpec.getDestination()).isReplaceTimeChunks()) {
+        return "msq_index_batch_replace";
```
I feel this is to easy to break. Its much better we add a label to the destination in case that is what you are looking for and then use destinaton.getLablel() here. 
WDYT ?

cryptoe replied:
I am unclear about how this would change in autocompaction.

cryptoe reviewed (commented):
I feel most of the changes should go into the core druid since all the task runners should be able to use this feature.

kfaraz commented:
__Summary__
IIUC, this PR seems to be implementing a task laning mechanism to better utilize available task slots. The lane that a particular task may be assigned to is determined by the new `label` field. Through configs, we may control the width and priority of each lane.

__My current thoughts on the design__
1. Since task laning may have applications outside of the contrib K8s extension too, the code changes should live in core Druid as suggested by @cryptoe . Another (less important) reason for this to be in core is the fact that we are adding a method to the `Task` API which currently has its only use in a contrib extension.

2. Instead of the new `getLabel` method in `Task`, we should have something more like `getTaskLaneInfo` that returns a `TaskLaneInfo` object which may be augmented later if needed. It should also be possible to manipulate the details of the lane info (i.e. the label in this case) using the task context. I guess the current problem in scheduling arises from the fact that we are limited to using immutable fields such as `type` and `datasource` to determine how a task should be scheduled. Powering this through the task context will allow users much more flexibility.

3. @YongGang , I suppose it would have been nice to put up a proposal for this design first before starting with the implementation. But since we already have the PR, please include details of the current design and any alternatives in the PR description.

kfaraz commented:
Marked this as a `Design Review` to get some more eyes on this.

YongGang commented:
@kfaraz thanks for your comment, it makes sense to me. 
I can move the task lane info/config to core Druid but the task laning mechanism will only be implemented in contrib K8s extension in this PR, does it sound good to you?

YongGang commented:
> Instead of the new getLabel method in Task, we should have something more like getTaskLaneInfo that returns a TaskLaneInfo object which may be augmented later if needed. It should also be possible to manipulate the details of the lane info (i.e. the label in this case) using the task context. I guess the current problem in scheduling arises from the fact that we are limited to using immutable fields such as type and datasource (and maybe context parameter taskPriority) to determine how a task should be scheduled. Powering this through the task context will allow users much more flexibility.

It's a good idea. However, I'd like to emphasize that task labeling serves multiple purposes beyond scheduling. For instance, identifying the costs associated with querying from deep storage tasks can be hard now, as they fall under the same MSQ task type as others. With task labeling, we can include this information in task reports, facilitating detailed cost analysis. Therefore, I propose keeping the `getLabel` method separate, enabling its use in various contexts beyond scheduling alone.

YongGang commented:
Updated the PR in response to feedback:
1. Move task lane config classes to core Druid. `TaskLaneRegistry` remains in the extension code due to its ties with task runner capacity config, which vary across different task runners, making it unsuitable for universal use.
2. The MSQ task label now derives from `MSQDestination`'s label, enhancing its adaptability to changes.
3. Modified `CliPeon` to demonstrate how task label is utilized in Peon monitoring and reporting.

@kfaraz @cryptoe

YongGang commented:
I have isolated the changes related to task labeling into a new PR https://github.com/apache/druid/pull/16041 to simplify the review process. I'm now converting this PR back to draft status to facilitate further Design Review discussions.
